src/main/java/appeng/fluids/parts/PartFluidStorageBus.java

```diff
@@ -0,0 +1,515 @@
+	}
+
+	@Override
+	public void readFromNBT( final NBTTagCompound data )
+	{
+		super.readFromNBT( data );
+		this.config.readFromNBT( data, "config" );
+	}
+
+	@Override
+	public void writeToNBT( final NBTTagCompound data )
+	{
+		super.writeToNBT( data );
+		this.config.writeToNBT( data, "config" );
+	}
+
+	@Override
+	public boolean isValid( final Object verificationToken )
+	{
+		return this.handler == verificationToken;
+	}
+
+	@Override
+	public void postChange( final IBaseMonitor<IAEFluidStack> monitor, final Iterable<IAEFluidStack> change, final IActionSource source )
+	{
+		try
+		{
+			if( this.getProxy().isActive() )
+			{
+				this.getProxy().getStorage().postAlterationOfStoredItems( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ), change, this.source );
+			}
+		}
+		catch( final GridAccessException e )
+		{
+			// :(
+		}
+	}
+
+	public MEInventoryHandler<IAEFluidStack> getInternalHandler()
+	{
+		if( this.cached )
+		{
+			return this.handler;
+		}
+
+		final boolean wasSleeping = this.monitor == null;
+
+		this.cached = true;
+		final TileEntity self = this.getHost().getTile();
+		final TileEntity target = self.getWorld().getTileEntity( self.getPos().offset( this.getSide().getFacing() ) );
+		final int newHandlerHash = this.createHandlerHash( target );
+
+		if( newHandlerHash != 0 && newHandlerHash == this.handlerHash )
+		{
+			return this.handler;
+		}
+
+		this.handlerHash = newHandlerHash;
+		this.handler = null;
+		this.monitor = null;
+		if( target != null )
+		{
+			IMEInventory<IAEFluidStack> inv = this.getInventoryWrapper( target );
+			//TODO: Do something here
+			/*if( inv instanceof MEMonitorIInventory )
+			{
+				final MEMonitorIInventory h = (MEMonitorIInventory) inv;
+				h.setMode( (StorageFilter) this.getConfigManager().getSetting( Settings.STORAGE_FILTER ) );
+			}*/
+
+			if( inv instanceof ITickingMonitor )
+			{
+				this.monitor = (ITickingMonitor) inv;
+				this.monitor.setActionSource( new MachineSource( this ) );
+			}
+
+			if( inv != null )
+			{
+				this.checkInterfaceVsStorageBus( target, this.getSide().getOpposite() );
+
+				this.handler = new MEInventoryHandler<IAEFluidStack>( inv, AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) );
+
+				this.handler.setBaseAccess( (AccessRestriction) this.getConfigManager().getSetting( Settings.ACCESS ) );
+				this.handler.setWhitelist( this.getInstalledUpgrades( Upgrades.INVERTER ) > 0 ? IncludeExclude.BLACKLIST : IncludeExclude.WHITELIST );
+				this.handler.setPriority( this.getPriority() );
+
+				final IItemList<IAEFluidStack> priorityList = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createList();
+
+				final int slotsToUse = 18 + this.getInstalledUpgrades( Upgrades.CAPACITY ) * 9;
+				for( int x = 0; x < this.config.getSlots() && x < slotsToUse; x++ )
+				{
+					final IAEItemStack is = this.config.getAEStackInSlot( x );
+					if( is != null )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)